[mlir] Report line number from file rather than split

jpienaar opened the pull request:
Add convention for lexer if the last file is contained in the first, then the first is used for error reporting. This requires that these two overlap to make it easy to find the corresponding spots. Enables going from

```
within split at mlir/test/IR/invalid.mlir::10 offset :6:9: error: reference to an undefined block
```

to

```
mlir/test/IR/invalid.mlir:15:9: error: reference to an undefined block
```

This does change the split to not produce always null terminated buffers and tools that need it to do so themselves (which is mostly by copying - this may have no actual impact as previously this was a copy too). Alternatively I could copy both and then split buffers are always null-terminated, but that seems worse.

joker-eph commented on mlir/include/mlir/Support/ToolUtilities.h:
```diff
@@ -45,6 +51,15 @@ splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer> originalBuffer,
                       llvm::StringRef outputSplitMarker = "");
+
+/// Same as above, but for case where the original buffer is not used in any
+/// form.
```
I'm not sure I understand the comment?

jpienaar replied:
Updated to be more explicit about the handler.

joker-eph commented on mlir/lib/Tools/mlir-translate/MlirTranslateMain.cpp:
```diff
@@ -138,6 +138,11 @@ LogicalResult mlir::mlirTranslateMain(int argc, char **argv,
+    // the buffer does not guarantee null-termination. Make a copy of the buffer
+    // to ensure null-termination.
+    ownedBuffer = llvm::MemoryBuffer::getMemBufferCopy(
+        ownedBuffer->getBuffer(), ownedBuffer->getBufferIdentifier());
```
Aren't we gonna make a copy even if there is no split?

jpienaar replied:
Good point, checked for null-termination before copying.

joker-eph reviewed (commented):
LG overall!

Can you split the lexer fixes out in a separate commit?

jpienaar committed:
[mlir] Report line number from file rather than split

Add convention for lexer if the last file is contained in the first,
then the first is used for error reporting. This requires that these two
overlap to make it easy to find the corresponding spots. Enables going
from

```
within split at mlir/test/IR/invalid.mlir::10 offset :6:9: error: reference to an undefined block
```

to

```
mlir/test/IR/invalid.mlir:15:9: error: reference to an undefined block
```

This does change the split to not produce always null terminated buffers and
tools that need it to do so themselves (which is mostly by copying - this may
have no actual impact as previously this was a copy too). Alternatively I could
copy both and then split buffers are always null-terminated, but that seems
worse.

hanhanW commented:
Hi, this commit breaks our downstream project [tests](https://github.com/iree-org/iree/blob/382c4fa7884acf1f50b85bd8f9beaa43167f4ee6/compiler/src/iree/compiler/Dialect/Util/Conversion/MemRefToUtil/test/memref_ops.mlir#L1-L22); I wonder if it is expected. Inline the snippet:

```mlir
// RUN: iree-opt --split-input-file \
// RUN:   --pass-pipeline='builtin.module(iree-util-test-conversion{widen-integers}, cse, canonicalize)' \
// RUN:   --verify-diagnostics %s | FileCheck %s

// -----
// Must be rank-0 or rank-1.
// expected-error @-3 {{conversion to util failed}}
util.func @verify_invalid_rank_2(%buffer: memref<4x2xf32>, %idx: index) -> f32{
  // expected-error @below {{failed to legalize operation 'memref.load'}}
  %0 = memref.load %buffer[%idx, %idx] : memref<4x2xf32>
  util.return %0 : f32
}

// -----
// Must have an identity map.
// expected-error @-3 {{conversion to util failed}}
#map = affine_map<(d0)[s0] -> (d0 * s0)>
util.func @verify_invalid_non_identity_map(%buffer: memref<4xf32, #map>, %idx: index) -> f32 {
  // expected-error @below {{failed to legalize operation 'memref.load'}}
  %0 = memref.load %buffer[%idx] : memref<4xf32, #map>
  util.return %0 : f32
}
```

If we delete all the `expected-error` check, the line numbers of `conversion to util failed` change; it can only be captured once.

Without the change, the error is

```
module {
}

// -----
within split at /tmp/z.mlir:5 offset :4:8: error: unexpected error: failed to legalize operation 'memref.load' that was explicitly marked illegal
  %0 = memref.load %buffer[%idx, %idx] : memref<4x2xf32>
       ^
within split at /tmp/z.mlir:5 offset :0:0: error: unexpected error: conversion to util failed
// -----
within split at /tmp/z.mlir:12 offset :5:8: error: unexpected error: failed to legalize operation 'memref.load' that was explicitly marked illegal
  %0 = memref.load %buffer[%idx] : memref<4xf32, #map>
       ^
within split at /tmp/z.mlir:12 offset :0:0: error: unexpected error: conversion to util failed
```

With the change, the error is:

```
module {
}

// -----
/tmp/z.mlir:8:8: error: unexpected error: failed to legalize operation 'memref.load' that was explicitly marked illegal
  %0 = memref.load %buffer[%idx, %idx] : memref<4x2xf32>
       ^
/tmp/z.mlir:0:0: error: unexpected error: conversion to util failed
// -----
/tmp/z.mlir:16:8: error: unexpected error: failed to legalize operation 'memref.load' that was explicitly marked illegal
  %0 = memref.load %buffer[%idx] : memref<4xf32, #map>
       ^
/tmp/z.mlir:0:0: error: unexpected error: conversion to util failed
```

The reporting number of both conversion failure becomes `0:0`. In this case, the tool is happy with capturing the failure once. I.e., I need to update the checks to

```mlir
// RUN: iree-opt --split-input-file \
// RUN:   --pass-pipeline='builtin.module(iree-util-test-conversion{widen-integers}, cse, canonicalize)' \
// RUN:   --verify-diagnostics %s | FileCheck %s

// -----
// Must be rank-0 or rank-1.
// expected-error @-7 {{conversion to util failed}}
util.func @verify_invalid_rank_2(%buffer: memref<4x2xf32>, %idx: index) -> f32{
  // expected-error @below {{failed to legalize operation 'memref.load'}}
  %0 = memref.load %buffer[%idx, %idx] : memref<4x2xf32>
  util.return %0 : f32
}

// -----
// Must have an identity map.
#map = affine_map<(d0)[s0] -> (d0 * s0)>
util.func @verify_invalid_non_identity_map(%buffer: memref<4xf32, #map>, %idx: index) -> f32 {
  // expected-error @below {{failed to legalize operation 'memref.load'}}
  %0 = memref.load %buffer[%idx] : memref<4xf32, #map>
  util.return %0 : f32
}
```

Is it the expected behavior after the change?

hanhanW commented:
Thanks for the suggestions! @jtuyls will take over our integration issue.

jpienaar commented:
Sounds good, we landed option 3 so potentially no further action needed your side.
